File: src/main/java/xyz/chunkstories/core/entity/EntityPlayer.java
//
// This file is a part of the Chunk Stories Core codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package xyz.chunkstories.core.entity;

import xyz.chunkstories.api.Location;
import xyz.chunkstories.api.entity.Controller;
import xyz.chunkstories.api.entity.Entity;
import xyz.chunkstories.api.entity.EntityDefinition;
import xyz.chunkstories.api.entity.traits.*;
import xyz.chunkstories.api.entity.traits.serializable.*;
import xyz.chunkstories.api.events.voxel.WorldModificationCause;
import xyz.chunkstories.api.graphics.MeshMaterial;
import xyz.chunkstories.api.gui.GuiDrawer;
import xyz.chunkstories.api.input.Input;
import xyz.chunkstories.api.item.inventory.InventoryHolder;
import xyz.chunkstories.api.item.inventory.ItemPile;
import xyz.chunkstories.api.player.Player;
import xyz.chunkstories.api.sound.SoundSource.Mode;
import xyz.chunkstories.api.util.ColorsTools;
import xyz.chunkstories.api.world.World;
import xyz.chunkstories.api.world.WorldMaster;
import xyz.chunkstories.core.entity.traits.TraitArmor;
import xyz.chunkstories.core.entity.traits.TraitFoodLevel;
import xyz.chunkstories.core.entity.traits.MinerTrait;
import xyz.chunkstories.core.entity.traits.TraitEyeLevel;
import xyz.chunkstories.core.entity.traits.TraitTakesFallDamage;
import org.joml.*;

import java.util.HashMap;

/**
 * Core/Vanilla player, has all the functionality you'd want from it:
 * creative/survival mode, flying and walking controller...
 */
public class EntityPlayer extends EntityHumanoid implements WorldModificationCause, InventoryHolder {
	private TraitControllable controllerComponent;

	protected TraitInventory inventory;
	TraitSelectedItem selectedItemComponent;

	private TraitName name;
	TraitCreativeMode creativeMode;
	TraitFlyingMode flyMode;

	TraitArmor armor;
	private TraitFoodLevel foodLevel;

	TraitVoxelSelection raytracer;

	private boolean onLadder = false;

	Location lastCameraLocation;

	int variant;

	public EntityPlayer(EntityDefinition t, World world) {
		super(t, world);

		//controllerComponent = new TraitController(this);
		inventory = new TraitInventory(this, 10, 4);
		selectedItemComponent = new TraitSelectedItem(this, inventory);
		name = new TraitName(this);
		creativeMode = new TraitCreativeMode(this);
		flyMode = new TraitFlyingMode(this);
		foodLevel = new TraitFoodLevel(this, 100);
		armor = new TraitArmor(this, 4, 1);

		raytracer = new TraitVoxelSelection(this) {

			@Override
			public Location getBlockLookingAt(boolean inside, boolean can_overwrite) {
				double eyePosition = stance.getStance().getEyeLevel();

				Vector3d initialPosition = new Vector3d(getLocation());
				initialPosition.add(new Vector3d(0, eyePosition, 0));

				Vector3d direction = new Vector3d(entityRotation.getDirectionLookingAt());

				if (inside)
					return EntityPlayer.this.world.getCollisionsManager().raytraceSelectable(new Location(EntityPlayer.this.world, initialPosition), direction,
							256.0);
				else
					return EntityPlayer.this.world.getCollisionsManager().raytraceSolidOuter(new Location(EntityPlayer.this.world, initialPosition), direction,
							256.0);
			}

		};

		new TraitInteractible(this) {

			@Override
			public boolean handleInteraction(Entity entity, Input input) {
				if (entityHealth.isDead() && input.getName().equals("mouse.right")) {

					Controller controller = controllerComponent.getController();//entity.traits.tryWith(TraitControllable.class, TraitControllable::getController);
					if (controller instanceof Player) {
						Player p = (Player) controller;
						p.openInventory(inventory);
						return true;
					}
				}
				return false;
			}
		};

		new PlayerOverlay(this);
		new TraitDontSave(this);

		new TraitEyeLevel(this) {

			@Override
			public double getEyeLevel() {
				return stance.getStance().getEyeLevel();
			}

		};

		new PlayerMovementController(this);
		controllerComponent = new EntityPlayerController(this);

		new MinerTrait(this);

		int variant = ColorsTools.getUniqueColorCode(this.getName()) % 6;
		HashMap<String, String> aaTchoum = new HashMap<>();
		aaTchoum.put("albedoTexture", "./models/human/variant" + variant + ".png");
		MeshMaterial customSkin = new MeshMaterial("playerSkin", aaTchoum, "opaque");
		new EntityHumanoidRenderer(this, customSkin);
	}

	// Server-side updating
	@Override
	public void tick() {

		// if(world instanceof WorldMaster)
		traits.with(MinerTrait.class, mt -> mt.tickTrait());

		// Tick item in hand if one such exists
		ItemPile pileSelected = this.traits.tryWith(TraitSelectedItem.class, eci -> eci.getSelectedItem());
		if (pileSelected != null)
			pileSelected.getItem().tickInHand(this, pileSelected);

		// Auto-pickups items on the ground
		if (world instanceof WorldMaster && (world.getTicksElapsed() % 60L) == 0L) {

			for (Entity e : world.getEntitiesInBox(getLocation(), new Vector3d(3.0))) {
				if (e instanceof EntityGroundItem && e.getLocation().distance(this.getLocation()) < 3.0f) {
					EntityGroundItem eg = (EntityGroundItem) e;
					if (!eg.canBePickedUpYet())
						continue;

					world.getSoundManager().playSoundEffect("sounds/item/pickup.ogg", Mode.NORMAL, getLocation(), 1.0f,
							1.0f);

					ItemPile pile = eg.getItemPile();
					if (pile != null) {
						ItemPile left = this.inventory.addItemPile(pile);
						if (left == null)
							world.removeEntity(eg);
						else
							eg.setItemPile(left);
					}
				}
			}
		}

		if (world instanceof WorldMaster) {
			// Food/health subsystem handled here decrease over time

			// Take damage when starving
			// TODO: move to trait
			if ((world.getTicksElapsed() % 100L) == 0L) {
				if (foodLevel.getValue() == 0)
					entityHealth.damage(TraitFoodLevel.HUNGER_DAMAGE_CAUSE, 1);
				else {
					// 27 minutes to start starving at 0.1 starveFactor
					// Takes 100hp / ( 0.6rtps * 0.1 hp/hit )

					// Starve slowly if inactive
					float starve = 0.03f;

					// Walking drains you
					if (this.entityVelocity.getVelocity().length() > 0.3) {
						starve = 0.06f;
						// Running is even worse
						if (this.entityVelocity.getVelocity().length() > 0.7)
							starve = 0.15f;
					}

					float newfoodLevel = foodLevel.getValue() - starve;
					foodLevel.setValue(newfoodLevel);
				}
			}

			// It restores hp
			// TODO move to trait
			if (foodLevel.getValue() > 20 && !entityHealth.isDead()) {
				if (entityHealth.getHealth() < entityHealth.getMaxHealth()) {
					entityHealth.setHealth(entityHealth.getHealth() + 0.01f);

					float newfoodLevel = foodLevel.getValue() - 0.01f;
					foodLevel.setValue(newfoodLevel);
				}
			}

			// Being on a ladder resets your jump height
			if (onLadder)
				traits.with(TraitTakesFallDamage.class, fd -> fd.resetFallDamage());

			// So does flying
			if (flyMode.get())
				traits.with(TraitTakesFallDamage.class, fd -> fd.resetFallDamage());
		}

		super.tick();

	}

	@Override
	public String getName() {
		return name.getName();
	}
}

File: src/main/java/xyz/chunkstories/core/entity/EntityPlayer.kt
//
// This file is a part of the Chunk Stories Core codebase
// Check out README.md for more information
// Website: http://chunkstories.xyz
//

package xyz.chunkstories.core.entity

import org.joml.Vector3d
import org.joml.Vector3dc
import xyz.chunkstories.api.Location
import xyz.chunkstories.api.entity.Entity
import xyz.chunkstories.api.entity.EntityDefinition
import xyz.chunkstories.api.entity.MeleeWeapon
import xyz.chunkstories.api.entity.traits.TraitDontSave
import xyz.chunkstories.api.entity.traits.TraitInteractible
import xyz.chunkstories.api.entity.traits.TraitSight
import xyz.chunkstories.api.entity.traits.serializable.*
import xyz.chunkstories.api.graphics.MeshMaterial
import xyz.chunkstories.api.gui.Layer
import xyz.chunkstories.api.gui.inventory.InventorySlot
import xyz.chunkstories.api.gui.inventory.InventoryManagementUIPanel
import xyz.chunkstories.api.input.Input
import xyz.chunkstories.api.item.inventory.Inventory
import xyz.chunkstories.api.item.inventory.InventoryOwner
import xyz.chunkstories.api.player.Player
import xyz.chunkstories.api.sound.SoundSource
import xyz.chunkstories.api.util.getUniqueColorCode
import xyz.chunkstories.api.world.World
import xyz.chunkstories.core.entity.traits.*
import java.util.*

/**
* Core/Vanilla player, has all the functionality you'd want from it:
* creative/survival mode, flying and walking controller...
*/
class EntityPlayer(t: EntityDefinition, world: World) : EntityHumanoid(t, world), InventoryOwner {
	private val controllerComponent: TraitControllable

	protected var traitInventory: TraitInventory
	internal var traitSelectedItem: TraitSelectedItem

	private val traitName: TraitName
	internal var traitCreativeMode: TraitCreativeMode
	internal var traitFlyingMode: TraitFlyingMode

	internal var traitArmor: TraitArmor
	private val traitFoodLevel: TraitFoodLevel

	internal var traitSight: TraitSight

	internal var lastCameraLocation: Location? = null
	internal var variant: Int = 0

	val name: String
		get() = traitName.name

	init {
		traitInventory = object : TraitInventory(this, 10, 4) {
			override fun createMainInventoryPanel(inventory: Inventory, layer: Layer): InventoryManagementUIPanel? {
				return inventory.run {
					//val crafts = loadRecipes(recipesTest, entity.world.content)
					val craftingStation = entity.traits[TraitCrafting::class]?.getCraftingStation()
					val craftingAreaSideSize = craftingStation?.craftingAreaSideSize ?: 0

					val ui = InventoryManagementUIPanel(layer, width * 20 + 16, height * 20 + 16 + 8 + 20 * 3 + 8 + 8)
					for (x in 0 until width) {
						for (y in 0 until height) {
							val slot = InventorySlot.RealSlot(this, x, y)
							val uiSlot = ui.InventorySlotUI(slot, x * 20 + 8, y * 20 + 8)
							ui.slots.add(uiSlot)
						}
					}

					entity.traits[TraitArmor::class]?.inventory?.let {
						for(armorSlot in 0 until it.width) {
							val slot = InventorySlot.RealSlot(it, armorSlot, 0)
							val uiSlot = ui.InventorySlotUI(slot, 0 * 20 + 8,  84 + armorSlot * 20 + 8)
							ui.slots.add(uiSlot)
						}
					}

					craftingStation?.bringUpCraftingMenuSlots(ui, 8 + height * 20 + 8)

					ui
				}
			}
		}
		traitSelectedItem = TraitSelectedItem(this, traitInventory)
		traitName = TraitName(this)
		traitCreativeMode = TraitCreativeMode(this)
		traitFlyingMode = TraitFlyingMode(this)
		traitFoodLevel = TraitFoodLevel(this, 100f)
		traitArmor = TraitArmor(this, 4, 1)

		traitSight =
				object : TraitSight(this) {
					override val headLocation: Location
						get() = Location(world, Vector3d(location).add(0.0, traitStance.stance.eyeLevel, 0.0))
					override val lookingAt: Vector3dc
						get() = traitRotation.directionLookingAt

				}

		val fists = object : MeleeWeapon {
			override val name = "fists"
			override val damage = 15f
			override val warmupMillis = 20
			override val cooldownMillis = 180
			override val reach = 5.0
			override val attackSound = "sounds/entities/human/punch_attack.ogg"
		}
		TraitMeleeCombat(this, fists)
		TraitHealthFoodOverlay(this)
		TraitDontSave(this)

		object : TraitEyeLevel(this) {
			override val eyeLevel: Double
				get() = traitStance.stance.eyeLevel

		}

		TraitMaybeFlyControlledMovement(this)
		controllerComponent = PlayerController(this)

		object : TraitInteractible(this) {
			override fun handleInteraction(entity: Entity, input: Input): Boolean {
				if (traitHealth.isDead && input.name == "mouse.right") {

					val controller = this@EntityPlayer.controller//entity.traits.tryWith(TraitControllable.class, TraitControllable::getController);
					if (controller is Player) {
						TODO("open inventory stuff again")
						// controller.openInventory(traitInventory.inventory)
						return true
					}
				}
				return false
			}
		}

		TraitMining(this)

		traitHealth =
				object : EntityHumanoidHealth(this) {
					override fun playDamageSound() {
						if (!isDead) {
							val i = 1 + Math.random().toInt() * 3
							entity.world.soundManager.playSoundEffect("sounds/entities/human/hurt$i.ogg", SoundSource.Mode.NORMAL, entity.location, Math.random().toFloat() * 0.4f + 0.8f, 5.0f)
						}
					}
				}

		TraitCrafting(this)
		TraitCanPickupItems(this)

		val variant = getUniqueColorCode(name) % 6
		val customSkin = MeshMaterial("playerSkin", mapOf("albedoTexture" to "./models/human/variant$variant.png"), "opaque")
		EntityHumanoidRenderer(this, customSkin)
		PlayerCamera(this)
	}
}